File: src/main/java/com/atsuishio/superbwarfare/recipe/ModPotionRecipes.java
package com.atsuishio.superbwarfare.recipe;

import com.atsuishio.superbwarfare.init.ModPotions;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.alchemy.Potion;
import net.minecraft.world.item.alchemy.PotionUtils;
import net.minecraft.world.item.alchemy.Potions;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class ModPotionRecipes {

    private static List<PotionRecipe> recipes = List.of();

    public static void register() {
        ItemStack water = potion(Potions.WATER);
        ItemStack shock = potion(ModPotions.SHOCK.get());
        ItemStack strongShock = potion(ModPotions.STRONG_SHOCK.get());
        ItemStack longShock = potion(ModPotions.LONG_SHOCK.get());
        recipes = List.of(
                new PotionRecipe(water, new ItemStack(Items.LIGHTNING_ROD), shock),
                new PotionRecipe(shock, new ItemStack(Items.GLOWSTONE_DUST), strongShock),
                new PotionRecipe(shock, new ItemStack(Items.REDSTONE), longShock)
        );
    }

    private static ItemStack potion(Potion potion) {
        return PotionUtils.setPotion(Items.POTION.getDefaultInstance(), potion);
    }

    public static boolean isIngredient(ItemStack stack) {
        return recipes.stream().anyMatch(recipe ->
                ItemStack.isSameItemSameTags(recipe.ingredient(), stack));
    }

    public static boolean hasMix(ItemStack input, ItemStack ingredient) {
        return findRecipe(input, ingredient) != null;
    }

    public static ItemStack mix(ItemStack input, ItemStack ingredient) {
        PotionRecipe recipe = findRecipe(input, ingredient);
        return recipe == null ? ItemStack.EMPTY : recipe.output().copy();
    }

    @Nullable
    private static PotionRecipe findRecipe(ItemStack input, ItemStack ingredient) {
        if (input.isEmpty() || input.getCount() != 1 || ingredient.isEmpty()) {
            return null;
        }

        for (PotionRecipe recipe : recipes) {
            if (ItemStack.isSameItemSameTags(recipe.input(), input)
                    && ItemStack.isSameItemSameTags(recipe.ingredient(), ingredient)) {
                return recipe;
            }
        }
        return null;
    }

    private record PotionRecipe(ItemStack input, ItemStack ingredient, ItemStack output) {
    }
}

File: src/main/kotlin/com/atsuishio/superbwarfare/recipe/ModPotionRecipes.kt
package com.atsuishio.superbwarfare.recipe

import com.atsuishio.superbwarfare.init.ModPotions
import com.atsuishio.superbwarfare.tools.isSameItemStack
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.alchemy.Potion
import net.minecraft.world.item.alchemy.PotionUtils
import net.minecraft.world.item.alchemy.Potions

object ModPotionRecipes {
    private var recipes = emptyList<PotionRecipe>()

    @JvmStatic
    fun register() {
        val water = potion(Potions.WATER)
        val shock = potion(ModPotions.SHOCK)
        val strongShock = potion(ModPotions.STRONG_SHOCK)
        val longShock = potion(ModPotions.LONG_SHOCK)
        recipes = listOf(
            PotionRecipe(water, ItemStack(Items.LIGHTNING_ROD), shock),
            PotionRecipe(shock, ItemStack(Items.GLOWSTONE_DUST), strongShock),
            PotionRecipe(shock, ItemStack(Items.REDSTONE), longShock)
        )
    }

    private fun potion(potion: Potion): ItemStack {
        return PotionUtils.setPotion(Items.POTION.defaultInstance, potion)
    }

    @JvmStatic
    fun isIngredient(stack: ItemStack): Boolean {
        return recipes.any { isSameItemStack(it.ingredient, stack) }
    }

    @JvmStatic
    fun hasMix(input: ItemStack, ingredient: ItemStack): Boolean {
        return findRecipe(input, ingredient) != null
    }

    @JvmStatic
    fun mix(input: ItemStack, ingredient: ItemStack): ItemStack {
        return findRecipe(input, ingredient)?.output?.copy() ?: ItemStack.EMPTY
    }

    private fun findRecipe(input: ItemStack, ingredient: ItemStack): PotionRecipe? {
        if (input.isEmpty || input.count != 1 || ingredient.isEmpty) return null
        return recipes.firstOrNull {
            isSameItemStack(it.input, input) && isSameItemStack(it.ingredient, ingredient)
        }
    }

    private data class PotionRecipe(val input: ItemStack, val ingredient: ItemStack, val output: ItemStack)
}
